compiler/frontend/src/org/jetbrains/kotlin/types/expressions/ExpressionTypingVisitorForStatements.java

```diff
@@ -441,6 +443,36 @@ protected KotlinTypeInfo visitAssignment(KtBinaryExpression expression, Expressi
                                   ? refineTypeByPropertyInType(bindingContext, leftOperand, leftType)
                                   : refineTypeFromPropertySetterIfPossible(bindingContext, leftOperand, leftType);
 
+
+        // Resolve assign overload
+        VariableDescriptor descriptor = BindingContextUtils.extractVariableFromResolvedCall(bindingContext, leftOperand);
+        OverloadResolutionResults<FunctionDescriptor> assignmentOperationDescriptors = new NameNotFoundResolutionResult<>();
+        KotlinType assignmentOperationType = null;
+        TemporaryTraceAndCache temporaryForAssignmentOperation = null;
+        boolean isDelegated = descriptor instanceof PropertyDescriptor && ((PropertyDescriptor) descriptor).isDelegated();
+        if (descriptor != null && !descriptor.isVar() && !isDelegated) {
+            ExpressionReceiver receiver = ExpressionReceiver.Companion.create(left, leftType, context.trace.getBindingContext());
+            temporaryForAssignmentOperation = TemporaryTraceAndCache.create(
+                    context, "trace to check assignment operation like '=' for", expression);
+            assignmentOperationDescriptors =
+                    components.callResolver.resolveBinaryCall(
```

Comment:
If there is no `assign` then rhs will be resolved twice: firstly in `temporaryForAssignmentOperation` trace, and then in regular one at lin 479. This is bad, because it will lead to potentially exponential time of analysis in case if rhs contains some code block (lambda, if, when, try), which also contains such assignment

Reply:
Acknowledged. I will have to dig a bit, to better understand this resolution enough, to resolve this issue.